### PR TITLE
TRIMs Support for Peloton

### PR DESCRIPTION
**Pull request by vittvolt**
This pr implements **trim/btrim/ltrim/rtrim** functions with llvm support for peloton.

Eaxample: 
create table t (str varchar(50));
select btrim(str, 'abc') from t;

The new StrWithLen type is stole from #905

### CODE REVIEW

**Comment by pmenon** [comment]
@vittvolt Can you rebase this?

File: CMakeLists.txt [hunk 1]
```diff
@@ -50,7 +50,7 @@ else()
     message(STATUS "The compiler ${CMAKE_CXX_COMPILER} has no C++1Y support. Please use a different C++ compiler.")
 endif()
 
-# Create a new pre-processor macro __PELOTONFILE__ that has a truncated 
+# Create a new pre-processor macro __PELOTONFILE__ that has a truncated
```
Comment (pmenon) [inline CMakeLists.txt]: Undo these changes.

File: src/codegen/type/varchar_type.cpp [hunk 2]
```diff
@@ -198,15 +289,18 @@ static std::vector<TypeSystem::ComparisonInfo> kComparisonTable = {
...
 // Binary operations
 static Like kLike;
+static BTrim kBTrim;
+static LTrim kLTrim;
+static RTrim kRTrim;
 static std::vector<TypeSystem::BinaryOpInfo> kBinaryOperatorTable = {
-    {OperatorId::Like, kLike}
-};
+    {OperatorId::Like, kLike}, {OperatorId::BTrim, kBTrim},
+    {OperatorId::LTrim, kLTrim}, {OperatorId::RTrim, kRTrim}};
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: Run clang-format.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -107,32 +107,32 @@ endforeach(test_src ${test_srcs})
...
 	    COMMAND ${perf_name}
-	)   
-	
+	)
+
 	# add test
 	add_test(${perf_name} ${CMAKE_BINARY_DIR}/test/${perf_name} --gtest_color=yes
            --gtest_output=xml:${CMAKE_BINARY_DIR}/test/${perf_name}.xml)
-	
+
```
Comment (pmenon) [inline test/CMakeLists.txt]: Undo changes to this file.

**Comment by vittvolt** [comment]
@pmenon should be good to merge now.

**Comment by apavlo** [comment]
@vittvolt I can merge this when the build passes.

@pmenon Let me know if you're okay with this.

File: src/codegen/type/varchar_type.cpp [hunk 4]
```diff
@@ -212,6 +212,97 @@ struct DateTrunc : public TypeSystem::BinaryOperatorHandleNull {
   }
 };
 
+struct Trim : public TypeSystem::UnaryOperator {
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: You'll need to subclass `UnaryOperatorHandleNull` here

File: src/codegen/type/varchar_type.cpp [hunk 5]
```diff
@@ -212,6 +212,97 @@ struct DateTrunc : public TypeSystem::BinaryOperatorHandleNull {
...
+    llvm::Value *ret = codegen.Call(StringFunctionsProxy::Trim,
+                                    {val.GetValue(), val.GetLength()});
+
+    llvm::Value *str_ptr = codegen->CreateExtractValue(ret, 0);
+    llvm::Value *str_len = codegen->CreateExtractValue(ret, 1);
+    return Value(Varchar::Instance(), str_ptr, str_len);
+  }
+};
+
+struct BTrim : public TypeSystem::BinaryOperator {
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: Subclass `BinaryOperatorHandleNull` in order to properly handle `NULL` inputs. This should be the case for all functions you implement here.

File: src/codegen/type/varchar_type.cpp [hunk 6]
```diff
@@ -231,14 +322,24 @@ static std::vector<TypeSystem::ComparisonInfo> kComparisonTable = {
...
+static BTrim kBTrim;
+static LTrim kLTrim;
+static RTrim kRTrim;
 static std::vector<TypeSystem::BinaryOpInfo> kBinaryOperatorTable = {
-    {OperatorId::Like, kLike}, {OperatorId::DateTrunc, kDateTrunc}};
+    {OperatorId::Like, kLike},
+    {OperatorId::DateTrunc, kDateTrunc},
+    {OperatorId::BTrim, kBTrim},
+    {OperatorId::LTrim, kLTrim},
+    {OperatorId::RTrim, kRTrim}};
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: Run `clang-format`

File: src/function/string_functions.cpp [hunk 7]
```diff
@@ -12,6 +12,7 @@
 
 #include <cctype>
 #include <string>
+#include <algorithm>
```
Comment (pmenon) [inline src/function/string_functions.cpp]: Sort includes alphabetically.

File: src/function/string_functions.cpp [hunk 8]
```diff
@@ -195,72 +196,113 @@ type::Value StringFunctions::Replace(const std::vector<type::Value> &args) {
   return (type::ValueFactory::GetVarcharValue(str));
 }
 
+StringFunctions::StrWithLen StringFunctions::LTrim(
```
Comment (pmenon) [inline src/function/string_functions.cpp]: `clang-format` this file.

File: src/codegen/type/varchar_type.cpp (hunk 6 above)
Comment (vittvolt) [reply src/codegen/type/varchar_type.cpp]: @pmenon This was the output after running the formatter.py script, any suggestions where to change?

Comment (pmenon) [reply src/codegen/type/varchar_type.cpp]: Ah, no change needed then. Wasn't sure if you'd run it already.
